flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -194,23 +193,28 @@ public boolean processInput() throws Exception {
 
 		int readingInputIndex = selectNextReadingInputIndex();
 		if (readingInputIndex == -1) {
-			return false;
+			return InputStatus.NOTHING_AVAILABLE;
```

Comment:
I think this might be not right? Shouldn't we return `END_OF_INPUT` in some cases here?

Reply:
I think it does not exist the case of return `END_OF_INPUT` here. Because if both inputs are finished, then `StreamTask` would call `context.allActionsCompleted()`, that means it would never call `inputProcessor.processInput()` any more. In other words, if `processInput` is called here that should mean at-least one input is not finished yet.

Actually I also felt this logic is difficult to understand. I guess the `readingInputIndex == -1` should only happen in the first call of `processInput`. In the following calls the `processInput` is triggered by available input, then the selected input index should never be -1. Is it right?

Reply:
After double checking the logic in `MailBoxProcessor#runMailboxLoop`, it has the optimization logic of warming hot path via 
```
if (!mailbox.hasMail()) {
	return true;
}
```
Then it would cause `selectNextReadingInputIndex` return -1 here, not only for the first call `processInput`. But it still does not exist the `END_OF_INPUT` case here.

Another concern is that are there any experiments to verify this hot path optimization has benefits in practice? I wonder it might waste some cpu time and also a bit delay the valid processing after available notification.